### PR TITLE
[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73528

Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

### CODE REVIEW

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

File: torch/ao/quantization/qconfig.py [hunk 1]
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
             qconfig = default_qat_qconfig_v2
     return qconfig
 
-def get_default_qconfig_dict(backend='fbgemm', version=0):
-    qconfig = get_default_qconfig(backend)
+def get_default_qconfig_dict_helper(qconfig, qconfig_transpose):
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:256]: nit: let's add an underscore to show this is not public facing

File: torch/ao/quantization/qconfig.py [hunk 2]
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
...
+                        (torch.nn.functional.conv2d, qconfig),
+                        (torch.nn.functional.conv3d, qconfig),
+                        (torch.nn.functional.conv_transpose1d, qconfig_transpose),
+                        (torch.nn.functional.conv_transpose2d, qconfig_transpose),
+                        (torch.nn.functional.conv_transpose3d, qconfig_transpose),
+                        (torch.nn.functional.linear, qconfig)]}
+
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend)
+    qconfig_transpose = qconfig
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:277]: maybe we can add a version argument for `get_default_qconfig` as well

File: torch/ao/quantization/qconfig.py [hunk 3]
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
...
+    if backend == "fbgemm":
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:278]: can we add a comment explaining why we need a different qconfig for convtranspose?

**Comment by jerryzh168** [comment]
Please add a Summary and Test for this PR

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

File: torch/ao/quantization/qconfig.py [hunk 4]
```diff
@@ -235,8 +239,8 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
                               weight=default_weight_fake_quant)
         else:
             qconfig = default_qat_qconfig
-    # Use the fused observer + fake_quant modules for doing QAT.
-    if version == 1:
+    # Use the fused observe + fake_quant modules for doing QAT.
+    elif version == 1:
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:243]: should we add a raise here as well?

Comment (jerryzh168) [reply torch/ao/quantization/qconfig.py:243]: sorry, missed the else below

File: torch/ao/quantization/qconfig.py [hunk 5]
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
...
+                        (torch.nn.functional.conv3d, qconfig),
+                        (torch.nn.functional.conv_transpose1d, qconfig_transpose),
+                        (torch.nn.functional.conv_transpose2d, qconfig_transpose),
+                        (torch.nn.functional.conv_transpose3d, qconfig_transpose),
+                        (torch.nn.functional.linear, qconfig)]}
+
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend, version)
+    qconfig_transpose = qconfig
+    # default_per_channel_weight_observer is not currently compatible with fbgemm backend
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:286]: we can add a link to this issue here as well I think: https://github.com/pytorch/pytorch/issues/47535

File: torch/ao/quantization/qconfig.py [hunk 6]
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
...
-        "": qconfig,
-        "object_type": [("reshape", default_reuse_input_qconfig)]
-    }
+    qconfig = get_default_qat_qconfig(backend, version)
+    qconfig_transpose = qconfig
+    # default_per_channel_weight_observer is not currently compatible with fbgemm backend
+    # so we have to modify the weight observer to MovingAverageMinMaxObserver or another
+    # per tensor supported observer
+    if backend == "fbgemm":
+        qconfig_transpose = QConfig(activation=qconfig.activation, weight=MovingAverageMinMaxObserver)
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:300]: we need a fake_quant for weight as well for qat, maybe change this to `default_weight_fake_quant`?

File: torch/ao/quantization/qconfig.py [hunk 7]
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
...
+                        (torch.nn.functional.linear, qconfig)]}
+
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend, version)
+    qconfig_transpose = qconfig
+    # default_per_channel_weight_observer is not currently compatible with fbgemm backend
+    # so we have to modify the weight observer to MovingAverageMinMaxObserver or another
+    # per tensor supported observer
+    if backend == "fbgemm":
+        qconfig_transpose = QConfig(activation=qconfig.activation, weight=MovingAverageMinMaxObserver)
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:290]: I think the weight observer here can be: `default_weight_observer`

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks! had a few inline comments

File: torch/ao/quantization/qconfig.py (hunk 5 above)
Comment (dzdang) [reply torch/ao/quantization/qconfig.py:286]: added

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

**Pull request closed without merging**